### PR TITLE
DROOLS-5277 DMN codegen inner composite, recursion, Any fix

### PR DESCRIPTION
**Pull request by tarilabs**

### CODE REVIEW

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java [hunk 1]
```diff
@@ -39,19 +47,90 @@ public DMNAllTypesIndex(DMNTypeSafePackageName.Factory packageName, DMNModel...
...
+            String simpleName = DMNDeclaredType.asJavaSimpleName(dmnType);
+            return mapNamespaceIndex.get(IndexKey.from(dmnType)).getPackageName().appendPackage(simpleName);
+        }
+        if (DMNTypeUtils.isFEELBuiltInType(dmnType)) {
+            return convertBuiltin(dmnType);
+        }
+        if (dmnType.getBaseType() == null) {
+            throw new IllegalStateException();
+        }
+        String baseConverted = converDMNTypeAsJava(dmnType.getBaseType());
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java]: typo `conver`

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java [hunk 2]
```diff
@@ -39,19 +47,90 @@ public DMNAllTypesIndex(DMNTypeSafePackageName.Factory packageName, DMNModel...
...
+                                .map(DMNModelTypesIndex.IndexValue::getPackageName);
     }
 
+    @Deprecated
     public boolean isIndexedClass(String typeName) {
         return namespaceOfClass(typeName).isPresent();
     }
+
+    public String asJava(DMNType fieldDMNType) {
+        String converted = converDMNTypeAsJava(fieldDMNType);
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java]: inline

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java [hunk 3]
```diff
@@ -39,19 +47,90 @@ public DMNAllTypesIndex(DMNTypeSafePackageName.Factory packageName, DMNModel...
...
+        return converted;
+    }
+
+    public String converDMNTypeAsJava(DMNType dmnType) {
+        if (mapNamespaceIndex.containsKey(IndexKey.from(dmnType))) {
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNAllTypesIndex.java]: .containsKey + get is a double lookup, can this be done wth one statement?

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java [hunk 4]
```diff
@@ -66,89 +60,16 @@ public String getOriginalMapKey() {
...
-        }
+        return index.asJava(fieldDMNType);
     }
 
     // This returns the generic type i.e. when Collection<String> then String
     private String fieldTypeUnwrapped() {
         if (fieldDMNType.isCollection()) {
-            String typeName = getBaseType(fieldDMNType);
-            return withPackage(typeName);
+            String typeName = index.asJava(fieldDMNType.getBaseType());
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java]: here we rather
1) inline the whole statement
2) split into significant different statements (`index.asJava` and `fieldDMNType.getBaseType`)

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java [hunk 5]
```diff
@@ -204,7 +127,11 @@ private boolean fieldIsBasic() {
     }
 
     private boolean fieldIsDifferentThanObject() {
-        return !fieldTypeUnwrapped().equals(OBJECT_TYPE);
+        if (DMNTypeUtils.isFEELAny(fieldDMNType)) {
+            return false;
+        }
+        boolean isOtherObject = fieldDMNType.isCollection() ? fieldTypeUnwrapped().equals(OBJECT_TYPE) : getObjectType().equals(OBJECT_TYPE);
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java:132]: rename to `isNotObjectType`

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNTypeUtils.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * Internal utility class
+ */
+public class DMNTypeUtils {
+
+    private DMNTypeUtils() {
+        // only static method in this class.
+    }
+
+    public static boolean isFEELBuiltInType(DMNType dmnType) {
+        return dmnType.getNamespace().equals(org.kie.dmn.model.v1_1.KieDMNModelInstrumentedBase.URI_FEEL) ||
```
Comment (lucamolteni) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNTypeUtils.java:34]: extract `dmnType.getNamespace()`

**Review by lucamolteni — APPROVED** [review]
Thank you, just minor corrections

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNTypeUtils.java (hunk 6 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNTypeUtils.java:34]: of all the comments, this one I don't see the benefit, that is I don't see the benefit to extract a method which invokes a getter.

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java (hunk 5 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/DMNDeclaredField.java:132]: ? this statement represent if it IS an `OBJECT_TYPE`.
line 134 returns the negated.

**Review by hellowdan — APPROVED** [review]

**Pull request merged**